pkg/controller/cleanup/cleanup_controller.go

```diff
@@ -0,0 +1,217 @@
+	}
+	opName, err := k8sutil.GetOperatorName()
+	if err != nil {
+		return err
+	}
+
+	// Only watch the deployment for this operator
+	p := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return e.Meta.GetName() == opName && e.Meta.GetNamespace() == opNs
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return e.Meta.GetName() == opName && e.Meta.GetNamespace() == opNs
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return e.MetaNew.GetName() == opName && e.MetaNew.GetNamespace() == opNs
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return false
+		},
+	}
+
+	// Watch for changes to primary resource Deployment
+	err = c.Watch(&source.Kind{Type: &appsv1.Deployment{}}, &handler.EnqueueRequestForObject{}, p)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileDeployment implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileDeployment{}
+
+// ReconcileDeployment reconciles a Deployment object
+type ReconcileDeployment struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+	// Namespace that the operator is watching
+	watchNs string
+}
+
+// ContainerJFRFinalizer is a finalizer to clean up external resources before exiting
+const ContainerJFRFinalizer = "rhjmc.redhat.com/containerjfrFinalizer"
+
+// Reconcile reads that state of the cluster for a Deployment object and makes changes based on the state read
+// and what is in the Deployment.Spec
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileDeployment) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	ctx := context.Background()
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling Operator Deployment")
+
+	// Fetch the Deployment instance
+	instance := &appsv1.Deployment{}
+	err := r.client.Get(ctx, request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// Check if this operator deployment is being deleted
+	if instance.GetDeletionTimestamp() != nil {
+		if hasCleanupFinalizer(instance) {
+			// Cleanly delete all recordings
+			err := r.deleteRecordings(ctx, r.watchNs)
+			if err != nil {
+				log.Error(err, "failed to delete recordings", "namespace", instance.Namespace,
+					"name", instance.Name)
```

Comment:
Should this return the `err` and restart the reconcile loop or is it meant to always fall through to removing the finalizer from here?

Reply:
Yes it should. Good catch!